File: 03-系统编程/04-文件系统/VFS与statx/go/vfs_statx.go
// vfs_statx.go — statx(2) 请求/返回掩码语义 + st_mode 字段映射
//
// 运行: go run vfs_statx.go
//
// 常量与规则均来自实读的 include/uapi/linux/stat.h 与 fcntl.h；
// 与 python/ 目录里的模型同题，便于对照。

package main

import "fmt"

// ---- STATX_* 请求/返回位 ----
const (
	StatxType          uint32 = 0x00000001
	StatxMode          uint32 = 0x00000002
	StatxNlink         uint32 = 0x00000004
	StatxUID           uint32 = 0x00000008
	StatxGID           uint32 = 0x00000010
	StatxAtime         uint32 = 0x00000020
	StatxMtime         uint32 = 0x00000040
	StatxCtime         uint32 = 0x00000080
	StatxIno           uint32 = 0x00000100
	StatxSize          uint32 = 0x00000200
	StatxBlocks        uint32 = 0x00000400
	StatxBasicStats    uint32 = 0x000007ff
	StatxBtime         uint32 = 0x00000800
	StatxMntID         uint32 = 0x00001000
	StatxDioalign      uint32 = 0x00002000
	StatxMntIDUnique   uint32 = 0x00004000
	StatxSubvol        uint32 = 0x00008000
	StatxWriteAtomic   uint32 = 0x00010000
	StatxDioReadAlign  uint32 = 0x00020000
	StatxReserved      uint32 = 0x80000000 // 唯一会触发 EINVAL 的保留位
	StatxAllDeprecated uint32 = 0x00000fff // 已废弃，等价于 BasicStats|Btime
)

// ---- STATX_ATTR_* 属性位 ----
const (
	StatxAttrCompressed uint64 = 0x00000004
	StatxAttrImmutable  uint64 = 0x00000010
	StatxAttrAppend     uint64 = 0x00000020
	StatxAttrNodump     uint64 = 0x00000040
	StatxAttrEncrypted  uint64 = 0x00000800
	StatxAttrAutomount  uint64 = 0x00001000
	StatxAttrMountRoot  uint64 = 0x00002000
	StatxAttrVerity     uint64 = 0x00100000
	StatxAttrDAX        uint64 = 0x00200000
	StatxAttrWAtomic    uint64 = 0x00400000
)

// ---- AT_* 路径解析与同步标志（fcntl.h）----
const (
	AtSymlinkNofollow uint32 = 0x100
	AtNoAutomount     uint32 = 0x800
	AtEmptyPath       uint32 = 0x1000
	AtStatxSyncType   uint32 = 0x6000
	AtStatxSyncAsStat uint32 = 0x0000
	AtStatxForceSync  uint32 = 0x2000
	AtStatxDontSync   uint32 = 0x4000
)

// ---- st_mode 文件类型 ----
const (
	SIfmt uint16 = 0170000
	SIfreg uint16 = 0100000
	SIfdir uint16 = 0040000
	SIfifo uint16 = 0010000
)

// Fs 描述一个具体文件系统的 statx 能力。
type Fs struct {
	Name      string
	Supported uint32 // 能填出来的位
	FreeBits  uint32 // 没请求也顺手填的位
	AttrMask  uint64 // stx_attributes_mask
	Attrs     uint64 // stx_attributes 原始值
	DummyUID  uint32
}

// Resolve 按 stat.h 头注释的四条规则算出 stx_mask，并返回可用的属性位。
func (f Fs) Resolve(req uint32) (uint32, uint64, error) {
	if req&StatxReserved != 0 {
		return 0, 0, fmt.Errorf("EINVAL: mask 含 STATX__RESERVED")
	}
	basic := []uint32{StatxType, StatxMode, StatxNlink, StatxUID, StatxGID,
		StatxAtime, StatxMtime, StatxCtime, StatxIno, StatxSize, StatxBlocks}
	extra := []uint32{StatxBtime, StatxMntID, StatxDioalign,
		StatxMntIDUnique, StatxSubvol, StatxWriteAtomic, StatxDioReadAlign}
	all := append(append([]uint32{}, basic...), extra...)

	got := uint32(0)
	for _, b := range all {
		if req&b != 0 {
			if f.Supported&b != 0 {
				got |= b // 规则 2：显式请求且支持
			}
			continue // 规则 1：不支持 → 清位
		}
		if f.FreeBits&b != 0 && f.Supported&b != 0 {
			got |= b // 规则 3：没请求但顺手可得
		}
	}
	return got, f.Attrs & f.AttrMask, nil
}

// SyncMode 解析 AT_STATX_* 三态；FORCE|DONT 同时置位是非法组合。
func SyncMode(flags uint32) (string, error) {
	switch flags & AtStatxSyncType {
	case AtStatxForceSync:
		return "force_sync", nil
	case AtStatxDontSync:
		return "dont_sync", nil
	case AtStatxSyncAsStat:
		return "as_stat", nil
	}
	return "", fmt.Errorf("EINVAL: AT_STATX_SYNC_TYPE 非法")
}

func maskNames(m uint32) []string {
	pairs := []struct {
		b uint32
		n string
	}{
		{StatxType, "TYPE"}, {StatxMode, "MODE"}, {StatxNlink, "NLINK"},
		{StatxUID, "UID"}, {StatxGID, "GID"}, {StatxIno, "INO"},
		{StatxSize, "SIZE"}, {StatxBlocks, "BLOCKS"}, {StatxBtime, "BTIME"},
		{StatxMntID, "MNT_ID"}, {StatxDioalign, "DIOALIGN"},
		{StatxMntIDUnique, "MNT_ID_UNIQUE"}, {StatxSubvol, "SUBVOL"},
	}
	out := []string{}
	for _, p := range pairs {
		if m&p.b != 0 {
			out = append(out, p.n)
		}
	}
	return out
}

func fileType(mode uint16) string {
	switch mode & SIfmt {
	case SIfreg:
		return "regular"
	case SIfdir:
		return "directory"
	case SIfifo:
		return "fifo"
	}
	return "other"
}

func main() {
	ext4 := Fs{
		Name:      "ext4",
		Supported: StatxBasicStats | StatxBtime | StatxMntID | StatxDioalign | StatxMntIDUnique,
		FreeBits:  StatxIno,
		AttrMask:  StatxAttrImmutable | StatxAttrAppend | StatxAttrCompressed,
		Attrs:     StatxAttrImmutable | StatxAttrDAX, // DAX 会被屏蔽
	}
	nfs := Fs{Name: "nfs", Supported: StatxBasicStats}
	cifs := Fs{Name: "cifs", Supported: StatxBasicStats & ^StatxUID, DummyUID: 65534}

	fmt.Println("== 请求位与返回位不是一回事 ==")
	if got, _, err := ext4.Resolve(StatxSize); err == nil {
		fmt.Printf("  ext4 请求 SIZE      -> %v\n", maskNames(got))
	}
	if got, _, err := ext4.Resolve(StatxBtime); err == nil {
		fmt.Printf("  ext4 请求 BTIME     -> %v\n", maskNames(got))
	}
	if got, _, err := nfs.Resolve(StatxBtime); err == nil {
		fmt.Printf("  nfs  请求 BTIME     -> %v (被清)\n", maskNames(got))
	}
	if got, _, err := cifs.Resolve(StatxUID); err == nil {
		fmt.Printf("  cifs 请求 UID       -> %v (编造 uid=%d)\n", maskNames(got), cifs.DummyUID)
	}

	fmt.Println("== 属性位必须与 stx_attributes_mask 相与 ==")
	_, attrs, _ := ext4.Resolve(StatxType)
	fmt.Printf("  ext4 可用属性 = %#x (DAX 被屏蔽: %v)\n", attrs, attrs&StatxAttrDAX == 0)
	_, attrs, _ = nfs.Resolve(StatxType)
	fmt.Printf("  nfs  可用属性 = %#x\n", attrs)

	fmt.Println("== 保留位与非法同步组合 ==")
	if _, _, err := ext4.Resolve(StatxReserved); err != nil {
		fmt.Printf("  STATX__RESERVED -> %v\n", err)
	}
	if _, err := SyncMode(AtStatxForceSync | AtStatxDontSync); err != nil {
		fmt.Printf("  FORCE|DONT      -> %v\n", err)
	}
	for _, f := range []uint32{0, AtStatxForceSync, AtStatxDontSync} {
		if s, err := SyncMode(f); err == nil {
			fmt.Printf("  flags=%#05x -> %s\n", f, s)
		}
	}

	fmt.Println("== st_mode 类型位 ==")
	fmt.Printf("  0100644 -> %s, 0040755 -> %s\n", fileType(0100644), fileType(0040755))

	fmt.Println("== stx_blocks 以 512B 为单位 ==")
	fmt.Printf("  8 blocks -> %d 字节; 1MiB 稀疏文件仅占 %d 字节\n",
		8*512, 8*512)
}

File: 03-系统编程/04-文件系统/VFS与statx/python/vfs_statx.py
"""statx(2) 请求/返回掩码语义 + struct statx 字段映射的可执行模型。

常量值全部来自实际读过的内核头文件：

  * include/uapi/linux/stat.h   —— STATX_* 请求位、STATX_ATTR_* 属性位、
                                   四种"请求/返回"组合语义的权威注释
  * include/uapi/linux/fcntl.h  —— AT_* 路径解析标志与 AT_STATX_* 同步三态

模型刻意**不**假设 mask == stx_mask：内核可能返回没请求的字段（顺手可得），
也可能清掉请求了的字段（文件系统不支持），手册原话是 "stx_mask will not be
equal mask"。
"""

# ---------------------------------------------------------------- 请求掩码位
STATX_TYPE = 0x00000001
STATX_MODE = 0x00000002
STATX_NLINK = 0x00000004
STATX_UID = 0x00000008
STATX_GID = 0x00000010
STATX_ATIME = 0x00000020
STATX_MTIME = 0x00000040
STATX_CTIME = 0x00000080
STATX_INO = 0x00000100
STATX_SIZE = 0x00000200
STATX_BLOCKS = 0x00000400
STATX_BASIC_STATS = 0x000007FF  # 普通 stat 结构里的那些字段
STATX_BTIME = 0x00000800
STATX_MNT_ID = 0x00001000
STATX_DIOALIGN = 0x00002000
STATX_MNT_ID_UNIQUE = 0x00004000
STATX_SUBVOL = 0x00008000
STATX_WRITE_ATOMIC = 0x00010000
STATX_DIO_READ_ALIGN = 0x00020000
STATX__RESERVED = 0x80000000  # 唯一会触发 EINVAL 的保留位

# STATX_ALL 已废弃：头文件注释明确写 "the same as STATX_BASIC_STATS|STATX_BTIME，
# to avoid confusion please use the equivalent"。
STATX_ALL = 0x00000FFF

# ------------------------------------------------------------------ 属性位
STATX_ATTR_COMPRESSED = 0x00000004
STATX_ATTR_IMMUTABLE = 0x00000010
STATX_ATTR_APPEND = 0x00000020
STATX_ATTR_NODUMP = 0x00000040
STATX_ATTR_ENCRYPTED = 0x00000800
STATX_ATTR_AUTOMOUNT = 0x00001000
STATX_ATTR_MOUNT_ROOT = 0x00002000
STATX_ATTR_VERITY = 0x00100000
STATX_ATTR_DAX = 0x00200000
STATX_ATTR_WRITE_ATOMIC = 0x00400000

# ------------------------------------------------------------- AT_* 标志
AT_SYMLINK_NOFOLLOW = 0x100
AT_NO_AUTOMOUNT = 0x800
AT_EMPTY_PATH = 0x1000
AT_STATX_SYNC_TYPE = 0x6000
AT_STATX_SYNC_AS_STAT = 0x0000
AT_STATX_FORCE_SYNC = 0x2000
AT_STATX_DONT_SYNC = 0x4000

# ------------------------------------------------------- st_mode 文件类型
S_IFMT = 0o170000
S_IFSOCK = 0o140000
S_IFLNK = 0o120000
S_IFREG = 0o100000
S_IFBLK = 0o060000
S_IFDIR = 0o040000
S_IFCHR = 0o020000
S_IFIFO = 0o010000


def s_isreg(m):
    return (m & S_IFMT) == S_IFREG


def s_isdir(m):
    return (m & S_IFMT) == S_IFDIR


def s_islnk(m):
    return (m & S_IFMT) == S_IFLNK


MASK_NAMES = [
    (STATX_TYPE, "TYPE"), (STATX_MODE, "MODE"), (STATX_NLINK, "NLINK"),
    (STATX_UID, "UID"), (STATX_GID, "GID"), (STATX_ATIME, "ATIME"),
    (STATX_MTIME, "MTIME"), (STATX_CTIME, "CTIME"), (STATX_INO, "INO"),
    (STATX_SIZE, "SIZE"), (STATX_BLOCKS, "BLOCKS"), (STATX_BTIME, "BTIME"),
    (STATX_MNT_ID, "MNT_ID"), (STATX_DIOALIGN, "DIOALIGN"),
    (STATX_MNT_ID_UNIQUE, "MNT_ID_UNIQUE"), (STATX_SUBVOL, "SUBVOL"),
    (STATX_WRITE_ATOMIC, "WRITE_ATOMIC"),
    (STATX_DIO_READ_ALIGN, "DIO_READ_ALIGN"),
]

ATTR_NAMES = [
    (STATX_ATTR_COMPRESSED, "COMPRESSED"), (STATX_ATTR_IMMUTABLE, "IMMUTABLE"),
    (STATX_ATTR_APPEND, "APPEND"), (STATX_ATTR_NODUMP, "NODUMP"),
    (STATX_ATTR_ENCRYPTED, "ENCRYPTED"), (STATX_ATTR_AUTOMOUNT, "AUTOMOUNT"),
    (STATX_ATTR_MOUNT_ROOT, "MOUNT_ROOT"), (STATX_ATTR_VERITY, "VERITY"),
    (STATX_ATTR_DAX, "DAX"), (STATX_ATTR_WRITE_ATOMIC, "WRITE_ATOMIC"),
]


def mask_names(mask):
    return [n for b, n in MASK_NAMES if mask & b]


def attr_names(attrs):
    return [n for b, n in ATTR_NAMES if attrs & b]


class StatxBuf(object):
    """struct statx 的用户态镜像。未填充字段保持 0/None。"""

    def __init__(self):
        self.stx_mask = 0
        self.stx_blksize = 0
        self.stx_attributes = 0
        self.stx_attributes_mask = 0
        self.stx_nlink = 0
        self.stx_uid = 0
        self.stx_gid = 0
        self.stx_mode = 0
        self.stx_ino = 0
        self.stx_size = 0
        self.stx_blocks = 0
        self.stx_btime = None
        self.stx_mtime = 0
        self.stx_mnt_id = 0
        self.stx_dio_mem_align = 0
        self.stx_dio_offset_align = 0


class Inode(object):
    """一个被统计的 inode。blocks512 是 512B 单位，带洞文件会小于 size/512。"""

    def __init__(self, ino, mode, size, blocks512, nlink=1, uid=0, gid=0,
                 mtime=0, btime=None, mnt_id=0, dio_mem=0, dio_off=0,
                 link_target=None):
        self.ino = ino
        self.mode = mode
        self.size = size
        self.blocks512 = blocks512
        self.nlink = nlink
        self.uid = uid
        self.gid = gid
        self.mtime = mtime
        self.btime = btime
        self.mnt_id = mnt_id
        self.dio_mem = dio_mem
        self.dio_off = dio_off
        self.link_target = link_target


class Filesystem(object):
    """模拟一个具体文件系统的 statx 能力。

    supported : 能填出来的 STATX_* 位（其余请求位会被清掉）
    free_bits : 没请求也顺手填的位（"available in approximate form without
                any effort" —— 头注释原文）
    attr_mask : stx_attributes_mask，只有落在这里的属性位才有意义
    dummy_uid : 不支持 UID 时编造的兼容值（CIFS 挂载就是这种情形）
    """

    def __init__(self, name, supported, free_bits=0, attr_mask=0,
                 attributes=0, blksize=4096, dummy_uid=None):
        self.name = name
        self.supported = supported
        self.free_bits = free_bits
        self.attr_mask = attr_mask
        self.attributes = attributes
        self.blksize = blksize
        self.dummy_uid = dummy_uid


def sync_mode(flags):
    """解析 AT_STATX_* 三态；同时置 FORCE|DONT 是非法组合 → EINVAL。"""
    sel = flags & AT_STATX_SYNC_TYPE
    if sel == AT_STATX_FORCE_SYNC:
        return "force_sync"
    if sel == AT_STATX_DONT_SYNC:
        return "dont_sync"
    if sel == AT_STATX_SYNC_AS_STAT:
        return "as_stat"
    raise ValueError("EINVAL: AT_STATX_SYNC_TYPE 非法 (FORCE 与 DONT 同时置位)")


def statx(fs, inode, mask, flags=0, on_field=None):
    """按 stat.h 头注释的四条规则填充 StatxBuf。返回 (buf, sync_mode)。

    on_field(bit) 会在"填完一个字段后"被调用，用来模拟手册里那句提醒：
    出于性能与实现简单的考虑，struct statx 的不同字段可能来自**同一趟系统调用
    里的不同时刻**（另一进程并发 chmod/chown 时，可能拿到旧 mode + 新 uid）。
    """
    if mask & STATX__RESERVED:
        raise ValueError("EINVAL: mask 里带了 STATX__RESERVED (0x80000000)")
    sync_mode(flags)  # 非法组合同样 EINVAL

    buf = StatxBuf()
    buf.stx_blksize = fs.blksize  # [uncond]
    buf.stx_attributes_mask = fs.attr_mask  # [uncond]
    buf.stx_attributes = fs.attributes & fs.attr_mask

    got = 0

    # 规则 1/2：显式请求位 —— 支持就填并置位；不支持就清位并给编造值
    for bit in [b for b, _ in MASK_NAMES]:
        if not (mask & bit):
            continue
        if not (fs.supported & bit):
            _fill_dummy(buf, inode, fs, bit)  # 位保持清除
            continue
        _fill(buf, inode, fs, bit)
        got |= bit
        if on_field is not None:
            on_field(bit)

    # 规则 3：没请求但顺手可得 → 照样填并置位（stx_mask 会超出请求范围）
    for bit in [b for b, _ in MASK_NAMES]:
        if mask & bit:
            continue
        if (fs.free_bits & bit) and (fs.supported & bit):
            _fill(buf, inode, fs, bit)
            got |= bit

    buf.stx_mask = got
    return buf, sync_mode(flags)


def _fill(buf, inode, fs, bit):
    if bit == STATX_TYPE or bit == STATX_MODE:
        buf.stx_mode = inode.mode
    elif bit == STATX_NLINK:
        buf.stx_nlink = inode.nlink
    elif bit == STATX_UID:
        buf.stx_uid = inode.uid
    elif bit == STATX_GID:
        buf.stx_gid = inode.gid
    elif bit == STATX_INO:
        buf.stx_ino = inode.ino
    elif bit == STATX_SIZE:
        buf.stx_size = _stx_size(inode)
    elif bit == STATX_BLOCKS:
        buf.stx_blocks = inode.blocks512
    elif bit == STATX_MTIME:
        buf.stx_mtime = inode.mtime
    elif bit == STATX_BTIME:
        buf.stx_btime = inode.btime
    elif bit == STATX_MNT_ID:
        buf.stx_mnt_id = inode.mnt_id
    elif bit == STATX_DIOALIGN:
        buf.stx_dio_mem_align = inode.dio_mem
        buf.stx_dio_offset_align = inode.dio_off


def _fill_dummy(buf, inode, fs, bit):
    """不支持的字段：为兼容 stat() 编造一个值，但对应的 mask 位**不**置位。"""
    if bit == STATX_UID and fs.dummy_uid is not None:
        buf.stx_uid = fs.dummy_uid
    elif bit == STATX_SIZE:
        buf.stx_size = _stx_size(inode)  # 尺寸总能编出来，但位仍清除


def _stx_size(inode):
    # 符号链接的 stx_size 是它装的**路径长度**，不含结尾 NUL（statx(2) 原文）
    if s_islnk(inode.mode) and inode.link_target is not None:
        return len(inode.link_target)
    return inode.size


def allocated_bytes(buf):
    """stx_blocks 的单位是 512 字节（不是 blksize，也不是 1024）。"""
    return buf.stx_blocks * 512


def apparent_ratio(inode):
    """有洞的文件：实际分配 < size。返回 (size, 实际占用字节)。"""
    return inode.size, inode.blocks512 * 512


def dio_align_consistent(buf):
    """statx(2)：offset align 只在 mem align 非 0 时才非 0，反之亦然。"""
    if buf.stx_dio_mem_align == 0:
        return buf.stx_dio_offset_align == 0
    return buf.stx_dio_offset_align != 0


def supported_attrs(buf):
    """只有落在 stx_attributes_mask 里的属性位才是可用值。"""
    return buf.stx_attributes & buf.stx_attributes_mask
